File: app/src/main/java/com/example/sendmessages/Service/MessageService.java
package com.example.sendmessages.Service;

import android.net.Uri;
import android.widget.EditText;
import android.widget.ImageView;

import androidx.annotation.Nullable;

import com.example.sendmessages.Adapters.RecyclerViewAdapterMessages;
import com.example.sendmessages.DTO.MessageDto;
import com.example.sendmessages.Entity.ChatsEntity;
import com.example.sendmessages.Entity.MessageEntity;
import com.example.sendmessages.Common.DataBase;
import com.example.sendmessages.Common.DateFormat;
import com.example.sendmessages.Mapping.MessageMapper;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.QuerySnapshot;

import java.time.LocalDate;
import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Класс отвечает за операции с БД,
 * в частности прием/отправка сообщений пользователю с которым ведется переписка.
 **/
public class MessageService {

    private FirebaseFirestore db;
    private RecyclerViewAdapterMessages adapterMessages;
    private Uri uriImage;
    private ImageView imageView;
    private EditText editText;
    private String usernameFrom, usernameToWhom;
    private MessageMapper mapper;
    private MessageEntity messageEntity;
    private ChatsEntity chatsEntityFrom, chatsEntityToWhom;
    private boolean boolChatExist = false;

    public MessageService() {
        mapper = new MessageMapper();
        db = FirebaseFirestore.getInstance();
    }

    public RecyclerViewAdapterMessages getAdapterMessages() {
        return adapterMessages;
    }

    public MessageService setAdapterMessages(RecyclerViewAdapterMessages adapterMessages) {
        this.adapterMessages = adapterMessages;
        return this;
    }

    public ImageView getImageView() {
        return imageView;
    }

    public MessageService setImageView(ImageView imageView) {
        this.imageView = imageView;
        return this;
    }

    public EditText getEditText() {
        return editText;
    }

    public MessageService setEditText(EditText editText) {
        this.editText = editText;
        return this;
    }

    public String getUsernameFrom() {
        return usernameFrom;
    }

    public MessageService setUsernameFrom(String usernameFrom) {
        this.usernameFrom = usernameFrom;
        return this;
    }

    public String getUsernameToWhom() {
        return usernameToWhom;
    }

    public MessageService setUsernameToWhom(String usernameToWhom) {
        this.usernameToWhom = usernameToWhom;
        return this;
    }

    public Uri getUriImage() {
        return uriImage;
    }

    public MessageService setUriImage(Uri uriImage) {
        this.uriImage = uriImage;
        return this;
    }

    public void deleteUriImage() {
        uriImage = null;
    }

    /**
     * Метод отвечает за создание сущности сообщения
     **/
    private boolean setMessagesEntity() {

        /** Проверка на наличие текста в текстовом поле **/
        boolean isExistText =
                !editText.getText().toString().trim().equals("")
                        && editText.getText().toString().trim().length() != 0;

        if (
                isExistText
                        || getUriImage() != null
        ) {
            messageEntity = new MessageEntity(
                    DateFormat.getFormatToDataBase().format(ZonedDateTime.now()),
                    usernameFrom,
                    editText.getText().toString().trim()
            );

            if (getUriImage() != null) {
                messageEntity.setUriImage(getUriImage().toString());
            }

            deleteUriImage();
            return true;
        } else {
            return false;
        }
    }


    /**
     * Метод отвечает за создание сущности чата, создается 2 сущности,
     * чтобы создать 2 одинаковые записи в бд с одинаковым id.
     * Чат А с В по id:1 и чат В с А по id:1
     **/
    private void setChatsEntity() {
        chatsEntityFrom = new ChatsEntity(
                usernameToWhom
        );
        chatsEntityToWhom = new ChatsEntity(
                chatsEntityFrom.getIdChats(),
                usernameFrom
        );
    }

    /**
     * Метод отвечает за добавление чатов в БД
     **/
    private void addChatsToDataBase() {

        chatsEntityFrom = new ChatsEntity(
                chatsEntityFrom.getIdChats(),
                chatsEntityFrom.getUsernameToWhom(),
                messageEntity.getMessage(),
                messageEntity.getDateTimeToDataBase()
        );
        chatsEntityToWhom = new ChatsEntity(
                chatsEntityToWhom.getIdChats(),
                chatsEntityToWhom.getUsernameToWhom(),
                messageEntity.getMessage(),
                messageEntity.getDateTimeToDataBase()
        );

        db
                .collection(DataBase.CHATS_DB)
                .document(usernameFrom)
                .collection(DataBase.ListTag.COLLECTIONS_CHATS_TAG)
                .document(usernameToWhom)
                .set(chatsEntityFrom);
        db
                .collection(DataBase.CHATS_DB)
                .document(usernameToWhom)
                .collection(DataBase.ListTag.COLLECTIONS_CHATS_TAG)
                .document(usernameFrom)
                .set(chatsEntityToWhom);
        boolChatExist = false;
    }

    /**
     * Метод отвечает за получение чата с конкретным пользователем из БД,
     * если его еще нет, то создается новый.
     * Так же выводит все сообщения между пользователями.
     **/
    public void getChat() {
        db
                .collection(DataBase.CHATS_DB)
                .document(usernameFrom)
                .collection(DataBase.ListTag.COLLECTIONS_CHATS_TAG)
                .document(usernameToWhom)
                .get()
                .addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                    @Override
                    public void onSuccess(DocumentSnapshot documentSnapshot) {
                        if (documentSnapshot.exists()) {
                            chatsEntityFrom = documentSnapshot
                                    .toObject(ChatsEntity.class);
                            boolChatExist = true;
                        } else {
                            setChatsEntity();
                        }
                        getMessagesFromDataBase();
                    }
                });
    }

    /**
     * Метод отвечает за обновление чата в БД, происходит перезапись значений:
     * Последнее сообщение, время отправки сообщения.
     **/
    private void updateChats() {

        db
                .collection(DataBase.CHATS_DB)
                .document(usernameFrom)
                .collection(DataBase.ListTag.COLLECTIONS_CHATS_TAG)
                .document(usernameToWhom)
                .update(
                        "lastMessage", messageEntity.getMessage(),
                        "timeMessageToDataBase", messageEntity.getDateTimeToDataBase()
                );
        db
                .collection(DataBase.CHATS_DB)
                .document(usernameToWhom)
                .collection(DataBase.ListTag.COLLECTIONS_CHATS_TAG)
                .document(usernameFrom)
                .update(
                        "lastMessage", messageEntity.getMessage(),
                        "timeMessageToDataBase", messageEntity.getDateTimeToDataBase()
                );
        boolChatExist = true;
    }

    /**
     * Метод отвечает за запись сообщения в БД.
     **/
    public void addMessagesToDataBase() {

        boolean booleanMess = setMessagesEntity();

        if (booleanMess) {
            if (!boolChatExist) {
                addChatsToDataBase();
            }
            db
                    .collection(DataBase.MESSAGES_DB)
                    .document(chatsEntityFrom.getIdChats().toString())
                    .collection(DataBase.ListTag.COLLECTIONS_MESSAGES_TAG)
                    .document(messageEntity.getDateTimeToDataBase())
                    .set(messageEntity);
            editText
                    .getText()
                    .clear();
            updateChats();
        }
    }

    /**
     * Метод отвечает за получение списка всех сообщений с пользователем с которым ведется переписка из БД.
     **/
    private void getMessagesFromDataBase() {
        List<MessageDto> messageList = new ArrayList<MessageDto>();

        try {
            db
                    .collection(DataBase.MESSAGES_DB)
                    .document(chatsEntityFrom.getIdChats().toString())
                    .collection(DataBase.ListTag.COLLECTIONS_MESSAGES_TAG)
                    .addSnapshotListener(
                            new EventListener<QuerySnapshot>() {
                                @Override
                                public void onEvent(
                                        @Nullable QuerySnapshot value,
                                        @Nullable FirebaseFirestoreException error
                                ) {
                                    LocalDate localDate = LocalDate.now();
                                    adapterMessages.deleteList();
                                    for (DocumentSnapshot ds : value.getDocuments()) {

                                        /**
                                         *  Проверка на существование даты, если ее нет,
                                         *  то конвертация через маппер не осуществляется.
                                         **/

                                        if (
                                                ds
                                                        .toObject(MessageEntity.class)
                                                        .getDateTimeToDataBase() != null
                                        ) {

                                            ZonedDateTime zonedDateTime =
                                                    ZonedDateTime.parse(
                                                            ds.toObject(MessageEntity.class).getDateTimeToDataBase(),
                                                            DateFormat.getFormatFromDataBase()
                                                    );

                                            /**
                                             *  Проверка на сегодняшнюю дату,
                                             *  если дата отправки сообщения совпадает с сегодняшней,
                                             *  то выводится только время, если нет,
                                             *  то дополнительно выводится год, месяц и день.
                                             **/

                                            if (zonedDateTime.toLocalDate().isEqual(localDate)) {
                                                mapper.setIsToday(false);
                                            } else {
                                                mapper.setIsToday(true);
                                            }

                                            MessageDto message = mapper
                                                    .getEntityToDto(ds.toObject(MessageEntity.class));
                                            messageList.add(message);
                                        }
                                        adapterMessages.setList(messageList);
                                    }
                                }
                            });

        } catch (Exception e) {

        }

    }
}

File: app/src/main/java/com/example/sendmessages/Service/MessageService.kt
package com.example.sendmessages.Service

import android.net.Uri
import android.widget.EditText
import android.widget.ImageView
import com.example.sendmessages.Adapters.RecyclerViewAdapterMessages
import com.example.sendmessages.Common.DataBase
import com.example.sendmessages.Common.DateFormat
import com.example.sendmessages.DTO.ChatsDto
import com.example.sendmessages.DTO.MessageDto
import com.example.sendmessages.Entity.ChatsEntity
import com.example.sendmessages.Entity.MessageEntity
import com.example.sendmessages.Interface.Mapping
import com.example.sendmessages.Mapping.ChatsMapper
import com.example.sendmessages.Mapping.MessageMapper
import com.example.sendmessages.Security.KeyGenerator
import com.example.sendmessages.Security.Keys
import com.example.sendmessages.Security.PrivateKeys
import com.example.sendmessages.Security.PublicKeys
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.EventListener
import com.google.firebase.firestore.FirebaseFirestore
import java.time.Instant
import java.time.LocalDate
import java.time.ZonedDateTime

/**
 * Класс отвечает за операции с БД,
 * в частности прием/отправка сообщений пользователю с которым ведется переписка.
 */
class MessageService() {
    private val db: FirebaseFirestore
    var adapterMessages: RecyclerViewAdapterMessages? = null
        private set
    var uriImage: Uri? = null
        private set
    var imageView: ImageView? = null
        private set
    var editText: EditText? = null
        private set
    var usernameFrom: String? = null
        private set
    var usernameToWhom: String? = null
        private set
    private val messageMapper: MessageMapper = MessageMapper()
    private val chatsMapper: ChatsMapper = ChatsMapper()
    private var messageDto: MessageDto? = null
    private var chatsDtoFrom: ChatsDto? = null
    private var chatsDtoToWhom: ChatsDto? = null
    private val keyGenerator = KeyGenerator()
    private var pairKeys: Pair<PublicKeys, PrivateKeys>? = null
    private var number: Long = 0
    private var boolChatExist = false

    init {
        db = FirebaseFirestore.getInstance()
    }

    fun setAdapterMessages(adapterMessages: RecyclerViewAdapterMessages?): MessageService {
        this.adapterMessages = adapterMessages
        return this
    }

    fun setImageView(imageView: ImageView?): MessageService {
        this.imageView = imageView
        return this
    }

    fun setEditText(editText: EditText?): MessageService {
        this.editText = editText
        return this
    }

    fun setUsernameFrom(usernameFrom: String?): MessageService {
        this.usernameFrom = usernameFrom
        return this
    }

    fun setUsernameToWhom(usernameToWhom: String?): MessageService {
        this.usernameToWhom = usernameToWhom
        return this
    }

    fun setUriImage(uriImage: Uri?): MessageService {
        this.uriImage = uriImage
        return this
    }

    fun deleteUriImage() {
        uriImage = null
    }

    /**
     * Метод отвечает за создание сущности сообщения
     */
    private fun setMessagesEntity(): Boolean {
        /** Проверка на наличие текста в текстовом поле  */
        val isExistText = (editText!!.text.toString().trim { it <= ' ' } != ""
            && editText!!.text.toString().trim { it <= ' ' }.length != 0)
        if ((isExistText
                || uriImage != null)) {
            messageDto = MessageDto(
                editText!!.text.toString().trim { it <= ' ' },
                DateFormat.getFormatToDataBase().format(ZonedDateTime.now()),
                (usernameFrom)!!,
                null
            )
            if (uriImage != null) {
                messageDto!!.uriImage = uriImage.toString()
            }
            deleteUriImage()
            return true
        } else {
            return false
        }
    }

    /**
     * Метод отвечает за создание сущности чата, создается 2 сущности,
     * чтобы создать 2 одинаковые записи в бд с одинаковым id.
     * Чат А с В по id:1 и чат В с А по id:1
     */
    private fun setChatsEntity() {
        chatsDtoFrom = ChatsDto(
            usernameToWhom = usernameToWhom!!
        )
        chatsDtoToWhom = ChatsDto(
            chatsDtoFrom!!.idChats,
            (usernameFrom)!!
        )
    }

    /**
     * Метод отвечает за добавление чатов в БД
     */
    private fun addChatsToDataBase() {
        number = keyGenerator.getNumberFromTwoLogins((usernameFrom)!!, (usernameToWhom)!!)
        pairKeys = keyGenerator.generateKeypair()
        val keys = keyGenerator.encryptKeys(pairKeys!!.first, pairKeys!!.second)
        val keysForDataBase = keyGenerator.encryptKeysForDataBase(keys, number)
        val lineKeys = keyGenerator.keysToString(keysForDataBase)

        chatsDtoFrom = ChatsDto(
            chatsDtoFrom!!.idChats,
            chatsDtoFrom!!.usernameToWhom,
            messageDto!!.message,
            messageDto!!.dateTimeToDataBase,
            lineKeys
        )

        chatsDtoToWhom = ChatsDto(
            chatsDtoToWhom!!.idChats,
            chatsDtoToWhom!!.usernameToWhom,
            messageDto!!.message,
            messageDto!!.dateTimeToDataBase,
            lineKeys
        )

        db
            .collection(DataBase.CHATS_DB)
            .document((usernameFrom)!!)
            .collection(DataBase.ListTag.COLLECTIONS_CHATS_TAG)
            .document((usernameToWhom)!!)
            .set(chatsMapper.getDtoToEntity(chatsDtoFrom, pairKeys))

        db
            .collection(DataBase.CHATS_DB)
            .document((usernameToWhom)!!)
            .collection(DataBase.ListTag.COLLECTIONS_CHATS_TAG)
            .document((usernameFrom)!!)
            .set(chatsMapper.getDtoToEntity(chatsDtoToWhom, pairKeys))

        boolChatExist = false
    }

    fun getChat() {
        /**
         * Метод отвечает за получение чата с конкретным пользователем из БД,
         * если его еще нет, то создается новый.
         * Так же выводит все сообщения между пользователями.
         */
        db
            .collection(DataBase.CHATS_DB)
            .document((usernameFrom)!!)
            .collection(DataBase.ListTag.COLLECTIONS_CHATS_TAG)
            .document((usernameToWhom)!!)
            .get()
            .addOnSuccessListener { documentSnapshot ->
                if (documentSnapshot.exists()) {
                    val chatsEntityFrom = documentSnapshot
                        .toObject(ChatsEntity::class.java)

                    val keys: Keys = keyGenerator.stringToKeys(chatsEntityFrom!!.lineKeys)
                    val number = keyGenerator.getNumberFromTwoLogins(usernameFrom!!, chatsEntityFrom.usernameToWhom)
                    val (keyConversion, publicKeys, privateKeys) = keyGenerator.decryptKeysForDataBase(keys, number)
                    pairKeys = keyGenerator.decryptKeys(keyConversion, publicKeys, privateKeys)

                    chatsDtoFrom = chatsEntityFrom.let { chatsMapper.getEntityToDto(it, pairKeys) }
                    boolChatExist = true
                } else {
                    setChatsEntity()
                }
                messagesFromDataBase
            }
    }

    /**
     * Метод отвечает за обновление чата в БД, происходит перезапись значений:
     * Последнее сообщение, время отправки сообщения.
     */
    private fun updateChats() {
        val message = keyGenerator.listBigIntegerToString(
            keyGenerator.encrypt(messageDto!!.message, pairKeys!!.first)
        )
        val dateTimeToDataBase = keyGenerator.listBigIntegerToString(
            keyGenerator.encrypt(messageDto!!.dateTimeToDataBase, pairKeys!!.first)
        )


        db
            .collection(DataBase.CHATS_DB)
            .document((usernameFrom)!!)
            .collection(DataBase.ListTag.COLLECTIONS_CHATS_TAG)
            .document((usernameToWhom)!!)
            .update(
                "lastMessage", message,
                "timeMessage", dateTimeToDataBase
            )
        db
            .collection(DataBase.CHATS_DB)
            .document((usernameToWhom)!!)
            .collection(DataBase.ListTag.COLLECTIONS_CHATS_TAG)
            .document((usernameFrom)!!)
            .update(
                "lastMessage", message,
                "timeMessage", dateTimeToDataBase
            )
        boolChatExist = true
    }

    /**
     * Метод отвечает за запись сообщения в БД.
     */
    fun addMessagesToDataBase() {
        val booleanMess = setMessagesEntity()
        if (booleanMess) {
            if (!boolChatExist) {
                addChatsToDataBase()
            }
            db
                .collection(DataBase.MESSAGES_DB)
                .document(chatsDtoFrom!!.idChats.toString())
                .collection(DataBase.ListTag.COLLECTIONS_MESSAGES_TAG)
                .document(Instant.now().epochSecond.toString())
                .set(messageMapper.getDtoToEntity(messageDto, pairKeys))

            editText?.text?.clear()
            updateChats()
        }
    }

    private val messagesFromDataBase: Unit
        /**
         * Метод отвечает за получение списка всех сообщений с пользователем с которым ведется переписка из БД.
         */
        private get() {
            val messageList: MutableList<MessageDto> = ArrayList()
            try {
                db
                    .collection(DataBase.MESSAGES_DB)
                    .document(chatsDtoFrom!!.idChats.toString())
                    .collection(DataBase.ListTag.COLLECTIONS_MESSAGES_TAG)
                    .addSnapshotListener(
                        EventListener { value, error ->
                            val localDate = LocalDate.now()
                            adapterMessages!!.deleteList()
                            for (ds: DocumentSnapshot in value!!.documents) {
                                /**
                                 * Проверка на существование даты, если ее нет,
                                 * то конвертация через маппер не осуществляется.
                                 */
                                if (
                                    ds.toObject<MessageEntity>(MessageEntity::class.java)
                                        ?.timeMessage != null
                                ) {

                                    val messageEntity = ds.toObject(MessageEntity::class.java)!!

                                    val zonedDateTime = ZonedDateTime.parse(
                                        keyGenerator.decrypt(
                                            keyGenerator.stringToListBigInteger(messageEntity.timeMessage),
                                            pairKeys!!.second
                                        ),
                                        DateFormat.getFormatFromDataBase()
                                    )
                                    /**
                                     * Проверка на сегодняшнюю дату,
                                     * если дата отправки сообщения совпадает с сегодняшней,
                                     * то выводится только время, если нет,
                                     * то дополнительно выводится год, месяц и день.
                                     */
                                    if (zonedDateTime.toLocalDate().isEqual(localDate)) {
                                        messageMapper.setIsToday(false)
                                    } else {
                                        messageMapper.setIsToday(true)
                                    }
                                    val message = messageMapper
                                        .getEntityToDto(messageEntity, pairKeys)
                                    messageList.add(message)
                                }
                            }
                            adapterMessages!!.setList(messageList)
                        })
            } catch (e: Exception) {
            }
        }
}